Source/WebDriver/soup/WebSocketServerSoup.cpp

```diff
@@ -0,0 +1,210 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebSocketServer.h"
+
+#include "CommandResult.h"
+#include <cstdio>
+#include <libsoup/soup-websocket-connection.h>
+#include <libsoup/soup.h>
+#include <optional>
+#include <tuple>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/text/WTFString.h>
+
+namespace WebDriver {
+
+static bool soupServerListen(SoupServer* server, const std::optional<String>& host, unsigned port, GError** error)
+{
+    static const auto options = static_cast<SoupServerListenOptions>(0);
+
+    if (!host || host.value() == "local"_s)
+        return soup_server_listen_local(server, port, options, error);
+
+    if (host.value() == "all"_s)
+        return soup_server_listen_all(server, port, options, error);
+
+    GRefPtr<GSocketAddress> address = adoptGRef(g_inet_socket_address_new_from_string(host.value().utf8().data(), port));
+    if (!address) {
+        g_set_error(error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "Invalid host IP address '%s'", host.value().utf8().data());
+        return false;
+    }
+
+    return soup_server_listen(server, address.get(), options, error);
+}
+
+static void handleIncomingHandshake(SoupServer* server, SoupServerMessage* message, const char* path, GHashTable*, gpointer userData)
+{
+    if (!soup_websocket_server_check_handshake(message, nullptr, nullptr, nullptr, nullptr)) {
+        soup_server_message_set_status(message, 500, "Invalid request");
+        return;
+    }
+    GRefPtr<SoupServerMessage> protectedMessage = message;
+    soup_server_pause_message(server, message);
+    auto* webSocketServer = static_cast<WebSocketServer*>(userData);
+    if (!webSocketServer) {
+        soup_server_message_set_status(message, 500, "Invalid server state");
+        return;
+    }
+
+    auto* requestBody = soup_server_message_get_request_body(message);
+    HTTPRequestHandler::Request handshakeMessage = {
+        String::fromUTF8(soup_server_message_get_method(message)),
+        String::fromUTF8(path),
+        requestBody->data,
+        static_cast<size_t>(requestBody->length)
+    };
+
+    webSocketServer->messageHandler().handleHandshake(WTFMove(handshakeMessage), [server, message](std::optional<HTTPRequestHandler::Response>&& response) {
+        if (!response) {
+            // No error found, follow with handshake procedure
+            soup_server_unpause_message(server, message);
+        } else {
```

Comment:
Nit: early return to allow dropping following block indentation.

Reply:
Replaced.